clang/lib/Sema/SemaDecl.cpp

```diff
@@ -15978,6 +15988,24 @@ Decl *Sema::ActOnFinishFunctionBody(Decl *dcl, Stmt *Body,
       CheckCoroutineWrapper(FD);
   }
 
+  // Diagnose invalid SYCL kernel entry point function declarations.
+  if (FD && !FD->isInvalidDecl() && !FD->isTemplated() &&
```

Comment:
Why `isTemplated`? All of these things look like rules you can check even on a dependent function, right? Also, 'isTemplated' just means one of its parents is at template, so in the case of a 'friend' this could actually be defined/declared even without the template being instantiated.

Reply:
Right, but `isTemplated` doesn't mean `isTemplateDecl`.  So this includes non-templates.

Reply:
It turned out to be pretty straight forward to enable diagnostics for dependent contexts. No additional changes were needed to diagnose coroutines. For deleted and defaulted functions, I had to duplicate the checks here in `SemaSYCL::CheckSYCLEntryPointFunctionDecl()`; the checks here handle non-dependent cases and the checks in `SemaSYCL::CheckSYCLEntryPointFunctionDecl()` handle dependent cases.